filebeat/input/filestream/internal/input-logfile/harvester_test.go

```diff
@@ -0,0 +1,331 @@
+
+	t.Run("assert already locked resource has to wait", func(t *testing.T) {
+		var wg sync.WaitGroup
+		mockHarvester := &mockHarvester{onRun: correctOnRun, wg: &wg}
+		hg := testDefaultHarvesterGroup(t, mockHarvester)
+		inputCtx := input.Context{Logger: logp.L(), Cancelation: context.Background()}
+
+		r, err := lock(inputCtx, hg.store, source.Name())
+		if err != nil {
+			t.Fatalf("cannot lock source")
+		}
+
+		gorountineChecker := newGoroutineChecker()
+
+		wg.Add(1)
+		hg.Start(inputCtx, source)
+
+		locked := true
+		gorountineChecker.waitWhileMoreGoroutinesWithFunc(func() {
+			// harvester is waiting to start
+			_, ok := hg.readers.table[source.Name()]
+			if ok && locked {
+				releaseResource(r)
+				locked = false
+			}
+		})
+
+		// wait until harvester.Run is done
+		wg.Wait()
+		// wait until goroutine that started `harvester.Run` is finished
+		gorountineChecker.waitOriginalCount()
+		require.Equal(t, 1, mockHarvester.runCount)
+		require.Nil(t, hg.StopGroup())
+	})
+
+	t.Run("assert already locked resource has no problem when harvestergroup is cancelled", func(t *testing.T) {
+		mockHarvester := &mockHarvester{onRun: correctOnRun}
+		hg := testDefaultHarvesterGroup(t, mockHarvester)
+		inputCtx := input.Context{Logger: logp.L(), Cancelation: context.Background()}
+		gorountineChecker := newGoroutineChecker()
+
+		r, err := lock(inputCtx, hg.store, source.Name())
+		if err != nil {
+			t.Fatalf("cannot lock source")
+		}
+		defer releaseResource(r)
+
+		hg.Start(inputCtx, source)
+
+		gorountineChecker.waitWhileMoreGoroutinesWithFunc(func() {
+			err := hg.StopGroup()
+			require.Error(t, err)
+		})
+
+		require.Equal(t, 0, mockHarvester.runCount)
+
+		gorountineChecker.waitOriginalCount()
+	})
+}
+
+func testDefaultHarvesterGroup(t *testing.T, mockHarvester Harvester) *defaultHarvesterGroup {
+	return &defaultHarvesterGroup{
+		readers:   newReaderGroup(),
+		pipeline:  &pipelinemock.MockPipelineConnector{},
+		harvester: mockHarvester,
+		store:     testOpenStore(t, "test", nil),
+		tg:        unison.TaskGroup{},
+	}
+}
+
+type gorountineChecker struct {
+	n    int
+	wait time.Duration
+}
+
+func newGoroutineChecker() *gorountineChecker {
+	return &gorountineChecker{
+		n:    runtime.NumGoroutine(),
+		wait: 10 * time.Millisecond,
+	}
+}
+
+func (c *gorountineChecker) waitWhileMoreGoroutinesWithFunc(f func()) {
+	for c.n < runtime.NumGoroutine() {
+		fmt.Println("waiting with function", c.n, runtime.NumGoroutine())
+		time.Sleep(c.wait)
+
+		f()
+	}
+}
+
+func (c *gorountineChecker) waitOriginalCount() {
+	fmt.Println("waiting until original", c.n, runtime.NumGoroutine())
+	for c.n < runtime.NumGoroutine() {
+		time.Sleep(c.wait)
+	}
+}
+
+type mockHarvester struct {
+	runCount int
+	wg       *sync.WaitGroup
+	onRun    func(input.Context, Source, Cursor, Publisher) error
+}
+
+func (m *mockHarvester) Run(ctx input.Context, s Source, c Cursor, p Publisher) error {
+	if m.wg != nil {
+		defer m.wg.Done()
+	}
+
+	m.runCount += 1
```

Comment:
looks like you have a race on `runCount`. The `Run` method and the test querying the count are run in different go-routines. Use `atomic.Int` or a mutex to prevent races.

Have you tested with `go test -race` ?

Reply:
I fixed all data races now.